txp/Pos/Txp/Toil/Types.hs

```diff
@@ -124,18 +127,43 @@ instance Default MemPool where
         }
 
 ----------------------------------------------------------------------------
--- ToilModifier
+-- UtxoModifier, UndoMap and AddrCoinsMap
 ----------------------------------------------------------------------------
 
 type UtxoModifier = MM.MapModifier TxIn TxOutAux
 type UndoMap = HashMap TxId TxUndo
-
-fromUtxo :: Utxo -> UtxoModifier
-fromUtxo = foldr (uncurry MM.insert) mempty . M.toList
+type AddrCoinMap = HashMap Address Coin
+
+utxoToModifier :: Utxo -> UtxoModifier
+utxoToModifier = foldr (uncurry MM.insert) mempty . M.toList
+
+-- | Takes utxo modifier and address-coin map with correspodning utxo
+-- and applies utxo modifier to map.
+-- Works for O(size of modifier * log (size of map)).
+applyUtxoModToAddrCoinMap
+    :: UtxoModifier
+    -> (AddrCoinMap, Utxo)
+    -> AddrCoinMap
+applyUtxoModToAddrCoinMap modifier (addrCoins, utxo) =
+    let outToPair = view _TxOut . toaOut in
+    let resolvedAddrs =
+          mapMaybe (fmap outToPair . flip M.lookup utxo)
+                   (MM.deletions modifier) in
+    let subAddress r c = if r < c then Just (c `unsafeSubCoin` r) else Nothing in
+    let updateHM (ad, coins) = HM.update (subAddress coins) ad in
+    let addrCoinsRest = foldr updateHM addrCoins resolvedAddrs in
+    -- We remove such TxIn which are already in wallet utxo.
+    let insertionsNotInUtxo = filter (not . flip M.member utxo . fst) (MM.insertions modifier) in
+    let modAddrCoins = map (outToPair . snd) insertionsNotInUtxo in
+    foldr (uncurry $ HM.insertWith unsafeAddCoin) addrCoinsRest modAddrCoins
```

Comment:
Without explicit types this function is hard to read.

Reply:
I think this code could be simplified by using `modifyHashMap`.

Reply:
It will be improved in #1735, I don't want to touch tested logic